### ISSUE TITLE
Errors uppercase/lowercase mismatch with service

### ISSUE THREAD

**Issue by gerchicov-bp** [OneDrive/onedrive-sdk-ios#112]
Now I get authentication error. But I can't handle it properly. Error info:
`Error Domain=com.microsoft.onedrive.errors Code=401 "unauthorized" UserInfo=0x196115d0 {NSLocalizedDescription=unauthorized, error= unauthenticated : Must be authenticated to use '/drive' syntax}`

And I should invent new methods how handle it properly because:
1)the error is about authentication. Why `-[NSError isAuthCanceledError]` returns `No`? Instead of this I somehow using messages in console need to guess that it is "client error" and handle its data...
2)Ok, let (1) is not a bug then how will you explain the following: `-[ODError code]` returns error in lowercase but `ODErrorCodes.m` contains error codes with both lowercase and uppercase letters? In other words in iOS 'a' != 'A'

**Comment by kevklam** [comment]
The authentication error is probably because of your modification to call the completion block immediately instead of waiting for the token request to complete. The auth provider does not have an OAuth token yet.

isAuthCanceledError returns YES if the user pressed 'cancel' - i.e. "is this error telling me that auth was canceled?". Also, 'authentication error' refers specifically to failures that occur during the authentication flow. Failing to append authentication headers to an API request is a client error ([NSError isClientError]).

Where is the exception being thrown from that is returning lowercase error codes?

**Comment by gerchicov-bp** [comment]
No, my modification fixes the incorrect order. Now it is exactly the following:
1)getToken
2)completion block (except of its dispach_async blocks)
3)dismissViewController

In your version (2) and (3) may be executed in any order and even simultaneously.

Sorry I meaned `isAuthenticationError`.

I don't know exactly how to reproduce that error - it just fails authentication through some time. But error code 401 and `-[ODError code]` returns "unauthenticated" but in `ODErrorCodes.m`:
`NSString *const ODUnAuthenticatedError = @"unAuthenticated";`

**Comment by kevklam** [comment]
Just to make sure I understand correctly, does the error tend to occur immediately after login or after the app has been running for some time? If the latter, would you mind checking the following:
- In ODAuthHelper.m, in +(BOOL)shouldRefreshSession - verify session.refreshToken is not nil
- Log [session.expires timeIntervalSinceNow] - when you get the 401, does this time difference look suspicious, like UTC/timezone difference?
- Is there any log message like "Refreshing session"?
